pkg/api/v20240812preview/openshiftcluster_validatestatic.go

```diff
@@ -412,3 +420,58 @@ func (sv openShiftClusterStaticValidator) validateDelta(oc, current *OpenShiftCl
 
 	return nil
 }
+
+func (sv openShiftClusterStaticValidator) validatePlatformWorkloadIdentityProfile(path string, pwip *PlatformWorkloadIdentityProfile) error {
+	// PlatformWorkloadIdentityProfile being empty is acceptable
+	if pwip == nil {
+		return nil
+	}
+
+	resourceIDPattern, _ := regexp.Compile("/subscriptions/.*/resourceGroups/.*/providers/.*/.*/.*")
+
+	// Validate the PlatformWorkloadIdentities
+	for n, p := range pwip.PlatformWorkloadIdentities {
+		// the regexp is statically defined and hard-coded, we shouldn't need to check for errors
+		match := resourceIDPattern.Match([]byte(p.ResourceID))
+		if !match {
+			return api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidParameter, fmt.Sprintf("%s.PlatformWorkloadIdentities[%d].resourceID", path, n), "ResourceID %s formatted incorrectly.", p.ResourceID)
+		}
+
+		if p.OperatorName == "" {
+			return api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidParameter, fmt.Sprintf("%s.PlatformWorkloadIdentities[%d].resourceID", path, n), "Operator name is empty.")
+		}
+
+		if strings.Split(p.ResourceID, "/")[7] != "userAssignedIdentities" {
+			return api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidParameter, fmt.Sprintf("%s.PlatformWorkloadIdentities[%d].resourceID", path, n), "Resource must be a user assigned identity.")
+		}
+	}
+	return nil
+}
+
+func (sv openShiftClusterStaticValidator) validatePlatformIdentities(oc *OpenShiftCluster) error {
+	pwip := oc.Properties.PlatformWorkloadIdentityProfile
+	spp := oc.Properties.ServicePrincipalProfile
+
+	if pwip == nil && spp == nil {
+		return api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidParameter, "properties.servicePrincipalProfile", "Must provide either an identity or service principal credentials.")
+	}
+
+	if pwip != nil && spp != nil && (spp.ClientID != "" || spp.ClientSecret != "") {
+		return api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidParameter, "properties.servicePrincipalProfile", "Cannot use identities and service principal credentials at the same time.")
+	}
+
+	clusterIdentityPresent := oc.Identity != nil
+	operatorRolePresent := false
+	if pwip != nil {
+		for _, pwi := range pwip.PlatformWorkloadIdentities {
+			if pwi.OperatorName != "" {
+				operatorRolePresent = true
+			}
+		}
+	}
+
+	if clusterIdentityPresent != operatorRolePresent {
+		return api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidParameter, "identity", "Cluster identity and operator roles require each other.")
+	}
```

Comment:
sorry for sporadic comments.
This is just a question, but in what case `operatorName` would be empty?

Reply:
I'm not sure, to be honest, but this scenario was called out as something to validate for. I'm happy to make changes to this validation if you have feedback though!

Reply:
We want to ensure that if a user passes in identities for an operator during a cluster create, the top-level cluster identity exists, otherwise the installation itself will fail because without the top-level cluster identity (`oc.Identity`) the RP won't be able to generate federated credentials for each operator identity.

This does raise an interesting case, though, which is the `az aro update`. With this validation present, I think that the user will have to also pass in the cluster identity along with it if providing a new identity to an existing cluster, even if the cluster identity details haven't changed.

@kimorris27 @tsatam do you have thoughts on how to best proceed?

Reply:
My reading of this conversation is that there are actually two different topics that ended up in this same thread, so I'll address them one at a time.

>This is just a question, but in what case operatorName would be empty?

In the context of this function, `validatePlatformIdentities` I think we can skip the check for empty `OperatorName` fields because we will be checking the same thing in `validatePlatformWorkloadIdentityProfile`. With that in mind, the section Ayato asked about:

```
operatorRolePresent := false
if pwip != nil {
	for _, pwi := range pwip.PlatformWorkloadIdentities {
		if pwi.OperatorName != "" {
			operatorRolePresent = true
		}
	}
}
```

can be replaced with `operatorRolePresent := pwip != nil` to simplify the code and avoid checking the same thing twice.

>This does raise an interesting case, though, which is the az aro update. With this validation present, I think that the user will have to also pass in the cluster identity along with it if providing a new identity to an existing cluster, even if the cluster identity details haven't changed.

>@kimorris27 @tsatam do you have thoughts on how to best proceed?

I think this code is fine as-is and we won't encounter this issue, since the RP responds to `PATCH` requests by getting the cluster doc from Cosmos DB and merging in any new fields the customer has provided prior to static validation. See https://github.com/Azure/ARO-RP/blob/master/pkg/frontend/openshiftcluster_putorpatch.go#L61-L172.

Reply:
I've pushed a commit to simply the code as suggested; thanks!